frontend/src/components/route/NewRouteMap.js

```diff
@@ -0,0 +1,187 @@
+                        <button className={"saveName"} onClick={saveRouteName}>Save route name</button>
+                    </div>
+                }
+                <GoogleMap mapContainerStyle={mapContainerStyle}
+                           zoom={10} center={centerMap()} options={options}
+                           onClick={clickHandler}>
+                    {waypoints.map(waypoint =>
+                        <Marker position={waypoint} key={waypoints.indexOf(waypoint)}
+                                label={labelMarker(waypoints, waypoint)}
+                        />)
+                    }
+                    <Polyline path={waypoints}/>
+                </GoogleMap>
+                <button className={"cancel"} onClick={redirectBackToRoutes}><MdCancel/>Cancel</button>
+                <button className={"done"} onClick={createRoute}><MdDone/>Done</button>
+            </MapContainer>
+        );
+    }
+
+    function centerMap() {
+        if (waypoints.length === 0) {
+            return {
+                lat: latitude,
+                lng: longitude,
+            }
+        }
+        return {
+            lat: waypoints[waypoints.length - 1].lat,
+            lng: waypoints[waypoints.length - 1].lng,
+        }
+    }
+
+    function clickHandler(event) {
+        setWaypoints([...waypoints, {lat: event.latLng.lat(), lng: event.latLng.lng()}]);
+    }
+
+    function labelMarker(collection, item) {
+        if ((collection.indexOf(item)) === 0) {
+            return "Start";
+        }
+        if ((collection.indexOf(item) + 1) === collection.length) {
+            return "End";
+        }
+        return String(collection.indexOf(item));
+    }
+
+    function saveRouteName() {
+        setShowName(!showName);
+    }
+
+    function redirectBackToRoutes() {
+        history.push("/routes");
+    }
+
+    function createRoute() {
+        const startWaypoint = {
+            startLatitude: waypoints[0].lat,
+            startLongitude: waypoints[0].lng,
+        }
+        const waypointsWithoutStart = waypoints.slice(1, waypoints.length);
+        const legs = [];
+
+        for (let i = 0; i < waypointsWithoutStart.length; i++) {
+            if (i === 0) {
+                legs[i] = {
+                    startLatitude: startWaypoint.startLatitude.toFixed(6),
+                    startLongitude: startWaypoint.startLongitude.toFixed(6),
+                    endLatitude: waypointsWithoutStart[i].lat.toFixed(6),
+                    endLongitude: waypointsWithoutStart[i].lng.toFixed(6),
+                }
+            }
+            if (i > 0) {
+                 legs[i] = {
+                    startLatitude: waypointsWithoutStart[i-1].lat.toFixed(6),
+                    startLongitude: waypointsWithoutStart[i-1].lng.toFixed(6),
+                    endLatitude: waypointsWithoutStart[i].lat.toFixed(6),
+                    endLongitude: waypointsWithoutStart[i].lng.toFixed(6),
+                }
+            }
+        }
```

Comment:
```suggestion
        const legs = waypointsWithoutStart.map((waypoint ,index)=> {
            if(index === 0){
                return {
                    startLatitude: startWaypoint.startLatitude.toFixed(6),
                    startLongitude: startWaypoint.startLongitude.toFixed(6),
                    endLatitude: waypointsWithoutStart[i].lat.toFixed(6),
                    endLongitude: waypointsWithoutStart[i].lng.toFixed(6),
                };
            }
            return  {
                    startLatitude: waypointsWithoutStart[i-1].lat.toFixed(6),
                    startLongitude: waypointsWithoutStart[i-1].lng.toFixed(6),
                    endLatitude: waypointsWithoutStart[i].lat.toFixed(6),
                    endLongitude: waypointsWithoutStart[i].lng.toFixed(6),
                }
            };
        });
```
functional style > imperative style :)